plugin/core/sessions.py

```diff
@@ -469,23 +470,62 @@ def can_start(cls, window: sublime.Window, initiating_view: sublime.View,
         """
         return None
 
+    @classmethod
+    def on_pre_start(cls, window: sublime.Window, initiating_view: sublime.View,
+                     workspace_folders: List[WorkspaceFolder], resolved: ResolvedStartupConfig) -> Optional[str]:
+        """
+        Callback invoked just before the language server subprocess is started. This is the place to do last-minute
+        adjustments to your "command" or "initializationOptions" in the passed-in "resolved" argument. You can also
+        choose to return a custom working directory, but consider that a language server shouldn't care about the
+        working directory.
+
+        :param      window:             The window
+        :param      initiating_view:    The initiating view
+        :param      workspace_folders:  The workspace folders
+        :param      resolved:           The resolved configuration
+
+        :returns:   A desired working directory, or None if you don't care
+        """
+        return None
```

Comment:
Does it allow to alter `workspace_folders`? I feel that if we would have a use-case for changing the workspace directory, we would as well want to set workspace folders to something else.

(I'm thinking of a use-case where we have to read some configuration from disk in which it would be defined what are the roots of the projects. But that would probably be even more complicated to get right...)

Reply:
I added more wording to signal that it's okay to change the workspace folders.

Reply:
It probably wouldn't work very well to change cwd as I imagine the server would still only be able to start once per window. But that's not important right now.